File: discord/slashCommands/checkRepos.go
package slashCommands

import (
	"github.com/bwmarrin/discordgo"
	"github.com/kjblanchard/sggDiscordBot/goonGithub"
	"log"
)

var (
	sysopRoleId = "907314874101665823"
)

func CheckIfUserInRole(roleToCheckFor string, roles []string) bool {
	for _, role := range roles {
		if role == roleToCheckFor {
			return true
		}
	}
	return false
}

func AddCheckReposSlashCommand(session *discordgo.Session, commands []*discordgo.ApplicationCommand) []*discordgo.ApplicationCommand {
	session.AddHandler(func(
		s *discordgo.Session,
		i *discordgo.InteractionCreate,
	) {
		data := i.ApplicationCommandData()
		if data.Name != "check-repos" {
			return
		}
		roles := i.Member.Roles
		if !CheckIfUserInRole(sysopRoleId, roles) {
			err := s.InteractionRespond(
				i.Interaction,
				&discordgo.InteractionResponse{
					Data: &discordgo.InteractionResponseData{
						Content: "Only Sysop can perform this action",
					},
				},
			)
			if err != nil {
				log.Print("Error responding! ", err)
			}
		} else {
			startNumber := 0
			if option, ok := getOptionValue(data.Options, "start-number"); ok {
				startNumber = int(option.IntValue())
			}
			err := s.InteractionRespond(
				i.Interaction,
				&discordgo.InteractionResponse{
					Type: discordgo.InteractionResponseDeferredChannelMessageWithSource,
				},
			)
			if err != nil {
				log.Print("Error defering", err)
			}
			go getReposAndRespondToInteraction(i.Interaction, session, startNumber)
		}
	})

	command := &discordgo.ApplicationCommand{
		Name:        "check-repos",
		Description: "Gets a list of 10 of the repos by kjb, starting at the first",
		Options: []*discordgo.ApplicationCommandOption{
			{
				Name:        "start-number",
				Description: "The number to start listing repositories from",
				Type:        discordgo.ApplicationCommandOptionInteger,
				Required:    false,
			},
		},
	}
	return append(commands, command)
}

func getOptionValue(options []*discordgo.ApplicationCommandInteractionDataOption, name string) (*discordgo.ApplicationCommandInteractionDataOption, bool) {
	for _, option := range options {
		if option.Name == name {
			return option, true
		}
	}
	return nil, false
}

func getReposAndRespondToInteraction(interaction *discordgo.Interaction, session *discordgo.Session, startNumber int) {
	repos := goonGithub.GetAllRepos()
	field := []*discordgo.MessageEmbedField{}
	start := startNumber
	end := start + 10

	for i := start; i < end; i++ {
		repo := repos[i]
		newField := &discordgo.MessageEmbedField{
			Name:  *repo.Name,
			Value: *repo.DefaultBranch,
		}
		field = append(field, newField)
	}
	// }
	content := "Heres a list of 10 the Repos!"
	embed := &discordgo.MessageEmbed{
		Title:  "Repos",
		Fields: field,
	}
	embeds := []*discordgo.MessageEmbed{embed}
	_, err := session.InteractionResponseEdit(interaction, &discordgo.WebhookEdit{
		Content: &content,
		Embeds:  &embeds,
	})
	if err != nil {
		log.Print("Error responding", err)
	}

}

File: python/discord_bot/slash_commands/check_repos.py
import logging

import discord
from discord import app_commands

from goon_github.repositories import get_all_repos

log = logging.getLogger(__name__)

SYSOP_ROLE_ID = 907314874101665823


def check_if_user_in_role(role_to_check_for: int, roles: list[discord.Role]) -> bool:
    for role in roles:
        if role.id == role_to_check_for:
            return True
    return False


def add_check_repos_slash_command(bot: discord.Client, guild: discord.Object):
    @bot.tree.command(name="check-repos", description="Gets a list of 10 of the repos by kjb, starting at the first", guild=guild)
    @app_commands.describe(start_number="The number to start listing repositories from")
    async def check_repos(interaction: discord.Interaction, start_number: int = 0):
        roles = interaction.user.roles
        if not check_if_user_in_role(SYSOP_ROLE_ID, roles):
            await interaction.response.send_message("Only Sysop can perform this action")
            return

        await interaction.response.defer()
        repos = get_all_repos()
        end = start_number + 10

        fields = []
        for i in range(start_number, min(end, len(repos))):
            repo = repos[i]
            fields.append(discord.EmbedField(name=repo.name, value=repo.default_branch or "main", inline=False))

        embed = discord.Embed(title="Repos")
        for field in fields:
            embed.add_field(name=field.name, value=field.value, inline=field.inline)

        await interaction.followup.send(content="Heres a list of 10 the Repos!", embed=embed)
